File: src/lib/models/z2m-device/z2mdevice.go
package z2mdevice

import (
	"encoding/json"
	"fmt"
)

/*
shared properties:

	type
	friendly_name
	disabled
	ieee_address
	interview_completed
	interviewing
	network_address
	supported

Router properties:

	definition {}
	manufacturer
	model_id
	power_source
	endpoints []
	software_build_id

EndDevice properties

	definition {}
	manufacturer
	model_id
	power_source
	endpoints []
	software_build_id (optional)
*/
type Z2mDevice struct {
	Type               string           `json:"type"`
	FriendlyName       string           `json:"friendly_name"`
	Disabled           bool             `json:"disabled"`
	IeeeAddress        string           `json:"ieee_address"`
	InterviewCompleted bool             `json:"interview_completed"`
	Interviewing       bool             `json:"interviewing"`
	NetworkAddress     int              `json:"network_address"`
	Supported          bool             `json:"supported"`
	Definition         *json.RawMessage `json:"definition"`
}

func ParseDevice(deviceData json.RawMessage) error {
	var rawDevice map[string]interface{}
	var z2mDevice Z2mDevice
	err := json.Unmarshal(deviceData, &rawDevice)
	if err != nil {
		return err
	}
	err = json.Unmarshal(deviceData, &z2mDevice)
	if err != nil {
		return err
	}

	hasDefiniton := z2mDevice.Definition != nil
	fmt.Printf("-- %s\n", rawDevice["type"])
	fmt.Printf("hasDefinition: %v\n", hasDefiniton)

	return nil
}

File: src/lib/models/z2m-device/z2m-device.ts

/* raw parse */

const z2m_device_types = [
  'Coordinator',
  'Router',
  'EndDevice',
] as const;

type Z2mDeviceEndpointBinding = {
  cluster: string;
  target: {
    /* only saw one type first-hand _*/
    type: 'endpoint';
    endpoint: number;
    ieee_address: string;
  };
} & {};
type Z2mDeviceEndpointConfiguredReporting = {
  attribute: string;
  cluster: string;
  maximum_reporting_interval: number;
  minimum_reporting_interval: number;
  reportable_change: number | number[];
} & {};

type Z2mDeviceEndpoint = {
  bindings: Z2mDeviceEndpointBinding;
  clusters: {
    input: string[];
    output: string[];
  };
  configured_reportings: Z2mDeviceEndpointConfiguredReporting[];
  scenes: {
    id: number;
    name: string;
  }[];
} & {};

type Z2mDeviceBase = {
  type: typeof z2m_device_types[number];
  friendly_name: string;
  disabled: boolean;
  ieee_address: string;
  interview_completed: boolean;
  interviewing: boolean;
  network_address: number;
  supported: boolean;
  // endpoints
  endpoints: Record<string, Z2mDeviceEndpoint>;
} & {};
